### PR TITLE
Support default spark run type for spark application

### PR DESCRIPTION
**Pull request by konjac**
This PR adds a new action allowing the user to specify the default spark application type from 3 options HDInsight, Cosmos Spark and Cosmos Serverless Spark.
There are two entries for this feature. One is in run context menu.
image

Another one is under tools menu.
image

After the user configures this, creating run configuration from project explorer will use the specified spark type. Also the "Submit spark application" on a file will automatically use the specified run config.

image

### CODE REVIEW

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/action/DefaultSparkApplicationTypeAction.kt [hunk 1]
```diff
@@ -0,0 +1,36 @@
+package com.microsoft.azure.hdinsight.spark.run.action
+
+import com.intellij.openapi.actionSystem.AnActionEvent
+import com.intellij.openapi.actionSystem.ToggleAction
+import com.microsoft.intellij.common.CommonConst
+import com.microsoft.tooling.msservices.components.DefaultLoader
+
+
+open class DefaultSparkApplicationTypeAction(private val sparkApplicationType : SparkApplicationType = SparkApplicationType.HDInsight)
+    : ToggleAction() {
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/action/DefaultSparkApplicationTypeAction.kt]: Could you combine `ToggleAction` with `AzureAnAction`?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/SparkBatchJobLocalRunConfigurationProducer.java [hunk 2]
```diff
@@ -59,12 +61,19 @@
...
+
+    public SparkBatchJobLocalRunConfigurationProducer(ConfigurationType configType, SparkApplicationType applicationType) {
+        super(configType);
+        this.applicationType = applicationType;
     }
 
     @Override
-    protected boolean setupConfigurationFromContext(LivySparkBatchJobRunConfiguration configuration, ConfigurationContext context, Ref<PsiElement> sourceElement) {
+    public boolean setupConfigurationFromContext(LivySparkBatchJobRunConfiguration configuration, ConfigurationContext context, Ref<PsiElement> sourceElement) {
+        if(DefaultSparkApplicationTypeAction.getSelectedSparkApplicationType() != this.applicationType) {
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/SparkBatchJobLocalRunConfigurationProducer.java]: Coding style

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/configuration/CosmosServerlessSparkConfigurationType.kt [hunk 3]
```diff
@@ -14,4 +16,11 @@ class CosmosServerlessSparkConfigurationType : CosmosSparkConfigurationType() {
     override fun getConfigurationFactories(): Array<ConfigurationFactory> {
         return arrayOf(CosmosServerlessSparkConfigurationFactory(this))
     }
+
+    companion object {
+        @JvmStatic
+        fun getInstance(): ConfigurationType {
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/configuration/CosmosServerlessSparkConfigurationType.kt]: Change `CosmosServerlessSparkConfigurationType` to object?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/configuration/CosmosSparkConfigurationType.kt [hunk 4]
```diff
@@ -48,4 +49,11 @@ open class CosmosSparkConfigurationType : ConfigurationType {
     override fun getConfigurationFactories(): Array<ConfigurationFactory> {
         return arrayOf(CosmosSparkConfigurationFactory(this))
     }
+
+    companion object {
+        @JvmStatic
+        fun getInstance() : ConfigurationType {
+            return ConfigurationTypeUtil.findConfigurationType(CosmosSparkConfigurationType::class.java)
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/configuration/CosmosSparkConfigurationType.kt]: Change the class to object for single instance?

**Comment by konjac** [comment]
New iteration allows users to leave the default run config type blank. Then a popup menu will prompt to let the user to choose which run config to use.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/configuration/CosmosSparkConfigurationType.kt [hunk 5]
```diff
@@ -30,11 +30,6 @@ import com.microsoft.intellij.util.PluginUtil
 import javax.swing.Icon
 
 open class CosmosSparkConfigurationType : ConfigurationType {
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/configuration/CosmosSparkConfigurationType.kt]: How about change it to `object` if it's singleton?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/configuration/LivySparkBatchJobRunConfigurationType.java [hunk 6]
```diff
@@ -65,5 +65,5 @@ public String getId() {
     @NotNull
     public static LivySparkBatchJobRunConfigurationType getInstance() {
         return ConfigurationTypeUtil.findConfigurationType(LivySparkBatchJobRunConfigurationType.class);
-    }
+}
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/configuration/LivySparkBatchJobRunConfigurationType.java]: White space change can be reverted.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/configuration/CosmosServerlessSparkConfigurationType.kt [hunk 7]
```diff
@@ -8,11 +8,6 @@ import com.microsoft.intellij.util.PluginUtil
 import javax.swing.Icon
 
 class CosmosServerlessSparkConfigurationType : ConfigurationType {
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/configuration/CosmosServerlessSparkConfigurationType.kt]: Also can be `object`

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/action/DefaultSparkApplicationTypeAction.kt [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+    companion object {
+        @JvmStatic
+        fun getSelectedSparkApplicationType() : SparkApplicationType? {
+            if (!DefaultLoader.getIdeHelper().isApplicationPropertySet(CommonConst.SPARK_APPLICATION_TYPE)) return SparkApplicationType.None
+            return SparkApplicationType.valueOf(DefaultLoader.getIdeHelper().getApplicationProperty(CommonConst.SPARK_APPLICATION_TYPE))
+        }
+    }
+
+    fun isSelected(): Boolean = this.sparkApplicationType == getSelectedSparkApplicationType()
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/action/DefaultSparkApplicationTypeAction.kt]: `getSelectedSparkApplicationType() != null && != None` ?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/actions/SparkSubmitJobAction.kt [hunk 9]
```diff
@@ -90,4 +105,18 @@ class SparkSubmitJobAction : AzureAnAction() {
...
     }
-}
\ No newline at end of file
+}
+
+open class LivySparkSubmitJobAction(val runConfigurationTypeInstance: ConfigurationType = LivySparkBatchJobRunConfigurationType.getInstance()) : SparkSubmitJobAction() {
+    override fun submitWithPopupMenu(anActionEvent: AnActionEvent): Boolean = false
+    override fun getRunConfigurationFromDataContext(dataContext: DataContext): RunnerAndConfigurationSettings? {
+        val configContext = ConfigurationContext.getFromContext(dataContext)
+        val factory = runConfigurationTypeInstance.configurationFactories[0]
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/actions/SparkSubmitJobAction.kt]: How about to assign `factory` with hardcoded type?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/actions/SparkSubmitJobAction.kt [hunk 10]
```diff
@@ -26,40 +26,55 @@ import com.intellij.execution.ExecutorRegistry
...
+        return if (DefaultSparkApplicationTypeAction.getSelectedSparkApplicationType() == SparkApplicationType.None) {
+            val action = ActionManagerEx.getInstance().getAction("SparkSubmitJobActionGroups") as QuickSwitchSchemeAction
+            action.actionPerformed(anActionEvent)
+            true
+        } else {
+            false
+        }
+    }
+    override fun onActionPerformed(anActionEvent: AnActionEvent?) {
+        if (anActionEvent == null) return
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/actions/SparkSubmitJobAction.kt]:
```suggestion
        if (anActionEvent == null) {
            return
        }
```

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/actions/SparkSubmitJobAction.kt [hunk 11]
```diff
@@ -26,40 +26,55 @@ import com.intellij.execution.ExecutorRegistry
...
+        if (submitWithPopupMenu(anActionEvent)) return
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/actions/SparkSubmitJobAction.kt]: Add braces for return

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/actions/SparkSubmitJobAction.kt [hunk 12]
```diff
@@ -26,40 +26,55 @@ import com.intellij.execution.ExecutorRegistry
...
+        val runConfigurationSetting = anActionEvent.dataContext.getData(RUN_CONFIGURATION_SETTING)
+                ?: getRunConfigurationFromDataContext(anActionEvent.dataContext) ?: return
         val clusterName = anActionEvent.dataContext.getData(CLUSTER)?.name
         val mainClassName = anActionEvent.dataContext.getData(MAIN_CLASS_NAME)
-
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/actions/SparkSubmitJobAction.kt]: Unnecessary change

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/actions/SparkSubmitJobAction.kt [hunk 13]
```diff
@@ -26,40 +26,55 @@ import com.intellij.execution.ExecutorRegistry
...
         submit(runConfigurationSetting, clusterName, mainClassName)
     }
 
     override fun update(event: AnActionEvent) {
         super.update(event)
-
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/actions/SparkSubmitJobAction.kt]: Unnecessary change

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/actions/SparkSubmitJobAction.kt [hunk 14]
```diff
@@ -90,4 +105,18 @@ class SparkSubmitJobAction : AzureAnAction() {
...
-}
\ No newline at end of file
+}
+
+open class LivySparkSubmitJobAction(val runConfigurationTypeInstance: ConfigurationType = LivySparkBatchJobRunConfigurationType.getInstance()) : SparkSubmitJobAction() {
+    override fun submitWithPopupMenu(anActionEvent: AnActionEvent): Boolean = false
+    override fun getRunConfigurationFromDataContext(dataContext: DataContext): RunnerAndConfigurationSettings? {
+        val configContext = ConfigurationContext.getFromContext(dataContext)
+        val factory = runConfigurationTypeInstance.configurationFactories[0]
+        val runConfig = factory.createTemplateConfiguration(configContext.project)
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/actions/SparkSubmitJobAction.kt]: Do we need to customize the config name with a pattern? We can find it next time to avoid creating lots of configs.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/action/DefaultSparkApplicationTypeAction.kt (hunk 8 above)
Comment (konjac) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/action/DefaultSparkApplicationTypeAction.kt]: This function is non-null. So null check is not necessary.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/action/SparkApplicationTypeOptionAction.kt [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+        val presentation = e.presentation
+        presentation.putClientProperty(Toggleable.SELECTED_PROPERTY, selected)
+        if (e.isFromContextMenu) {
+
+            presentation.icon = null
+        }
+    }
+}
+
+class SparkApplicationTypeOptionNone : SparkApplicationTypeOptionAction() {
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/action/SparkApplicationTypeOptionAction.kt]:
```suggestion
class SelectNoneSparkTypeAction : SparkApplicationTypeOptionAction() {
```

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/action/SparkApplicationTypeOptionAction.kt [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+    override fun getSparkApplicationType() : SparkApplicationType {
+        return SparkApplicationType.None
+    }
+}
+
+class SparkApplicationTypeOptionHDInsight : SparkApplicationTypeOptionAction() {
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/action/SparkApplicationTypeOptionAction.kt]:
```suggestion
class SelectHDInsightSparkTypeAction : SparkApplicationTypeOptionAction() {
```

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/action/SparkApplicationTypeOptionAction.kt [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+    override fun getSparkApplicationType() : SparkApplicationType {
+        return SparkApplicationType.HDInsight
+    }
+}
+
+class SparkApplicationTypeOptionCosmosSpark : SparkApplicationTypeOptionAction() {
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/action/SparkApplicationTypeOptionAction.kt]:
```suggestion
class SelectCosmosSparkTypeAction : SparkApplicationTypeOptionAction() {
```

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/action/SparkApplicationTypeOptionAction.kt [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+    override fun getSparkApplicationType() : SparkApplicationType {
+        return SparkApplicationType.CosmosSpark
+    }
+}
+
+class SparkApplicationTypeOptionCosmosServerlessSpark : SparkApplicationTypeOptionAction() {
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/action/SparkApplicationTypeOptionAction.kt]:
```suggestion
class SelectCosmosServerlessSparkTypeAction : SparkApplicationTypeOptionAction() {
```

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/actions/SparkSubmitJobAction.kt [hunk 19]
```diff
@@ -90,4 +106,18 @@ class SparkSubmitJobAction : AzureAnAction() {
...
     }
-}
\ No newline at end of file
+}
+
+open class SparkSelectTypeAndSubmitJobActionDelegate(private val sparkType: SparkApplicationType) : SparkApplicationTypeOptionAction() {
+    override fun getSparkApplicationType(): SparkApplicationType {
+        return this.sparkType
+    }
+    override fun runPostAction(e: AnActionEvent) {
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/actions/SparkSubmitJobAction.kt]: Can we just override `onActionPerformed()`?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/actions/SparkSubmitJobAction.kt [hunk 20]
```diff
@@ -90,4 +106,18 @@ class SparkSubmitJobAction : AzureAnAction() {
             model.isClusterSelectionEnabled = true
         }
     }
-}
\ No newline at end of file
+}
+
+open class SparkSelectTypeAndSubmitJobActionDelegate(private val sparkType: SparkApplicationType) : SparkApplicationTypeOptionAction() {
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/actions/SparkSubmitJobAction.kt]:
```suggestion
open class SparkSelectTypeAndSubmitAction(private val selectAction: SparkApplicationTypeOptionAction) by selectAction {

    override fun onActionPerformed() {
        selectAction.onActionPerformed()
        val submitAction = ActionManagerEx.getInstance().getAction("Actions.SubmitSparkApplicationAction")
        submitAction?.actionPerformed(e) ?: log().error("Can't find the action Actions.SubmitSparkApplicationAction to submit Spark application")
```

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/actions/SparkSubmitJobAction.kt [hunk 21]
```diff
@@ -90,4 +106,18 @@ class SparkSubmitJobAction : AzureAnAction() {
...
+    override fun getSparkApplicationType(): SparkApplicationType {
+        return this.sparkType
+    }
+    override fun runPostAction(e: AnActionEvent) {
+        val action = ActionManagerEx.getInstance().getAction("Actions.SubmitSparkApplicationAction")
+        action?.actionPerformed(e)
+    }
+}
+
+class LivySparkSubmitJobAction : SparkSelectTypeAndSubmitJobActionDelegate(SparkApplicationType.HDInsight)
```
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/actions/SparkSubmitJobAction.kt]: Can we do it like this:

```suggestion
class LivySparkSelectAndSubmitAction : SparkSelectTypeAndSubmitAction(ActionManagerEx.getInstance().getAction("SetDefaultSparkRunConfigurationTypeHDInsight"))
```
